### PR TITLE
ft(S3C-4383): Add getAccountIds

### PR DESCRIPTION
**Pull request by tmacro**

### CODE REVIEW

**Review by naren-rajendran — APPROVED** [review]

File: tests/unit/getAccountIds.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    }
+    if (inputArray.indexOf(canIdBad) !== -1) {
+        res.writeHead(errors.InvalidParameterValue.code);
+        return res.end(JSON.stringify(errors.InvalidParameterValue));
+    }
+    const output = [];
+    inputArray.forEach(canId => output.push({
+        id: serverDB[canId],
+        canId,
+    }));
```
Comment (miniscruff) [inline tests/unit/getAccountIds.js]: forEach + push can be map

```suggestion
    const output = inputArray.map(canId => {
        return {
            id: serverDB[canId],
            canId,
        }
    });
```

File: tests/unit/getAccountIds.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+function handler(req, res) {
+    const index = req.url.indexOf('?');
+    const data = querystring.parse(req.url.substring(index + 1));
+    let inputArray = data.canonicalIds;
+    if (!Array.isArray(inputArray)) {
+        inputArray = [inputArray];
+    }
+    if (inputArray.indexOf(canIdBad) !== -1) {
+        res.writeHead(errors.InvalidParameterValue.code);
```
Comment (jonathan-gramain) [inline tests/unit/getAccountIds.js:38]: Can you confirm that this is the actual behavior of Vault, i.e. it returns this error whenever **one of** the canonical IDs asked for does not exist? I would have expected that error would appear in the response array, only for the missing canonical ID entry.

**Review by jonathan-gramain — APPROVED** [review]

**Review by anurag4DSB — COMMENTED** [review]
I would suggest not to merge this PR, till we have a PR with respective functional tests in Vault for the same method. That way it might help you identify behavioral unit changes needed in Vaultclient w.r.t. Vault.

That will also help in verifying Jonathan's comment.

PS: I know vaultclient dependency needs to be merged first :)

File: tests/unit/getAccountIds.js (hunk 2 above)
Comment (tmacro) [reply tests/unit/getAccountIds.js:38]: This was only to force the mock to throw an error to test whether it is passed through correctly. As you point out a single error wouldn't cause the request to fail.

**Comment by tmacro** [comment]
Tests for the vault route are [https://github.com/scality/Vault/blob/f1b0d77aec5e74f39dee60fbc5ac1d34b028adf5/tests/functional/200-server-and-client/Accounts/get.js#L29](here) since `getAccountIds` is just wrapping the `getAccounts` method

File: lib/IAMClient.js [hunk 3]
```diff
@@ -673,6 +673,38 @@ class VaultClient {
         }, data, options.reqUid, null, options.logger);
     }
 
+    /**
+     * A getter for accountIds given a list of canonicalIDs
+     * @param{String[]} canonicalIds - list of canonicalIDs
```
Comment (miniscruff) [inline lib/IAMClient.js]:
```suggestion
     * @param {String[]} canonicalIds - list of canonicalIDs
```

File: lib/IAMClient.js [hunk 4]
```diff
@@ -673,6 +673,38 @@ class VaultClient {
...
+     * @param {Object} options - additional arguments
+     * @param {string} options.reqUid - the request UID
+     * @param {werelogs.RequestLogger} [options.logger] - Logger instance
+     * @param{Function} callback - the callback handling the response object
```
Comment (miniscruff) [inline lib/IAMClient.js]:
```suggestion
     * @param {Function} callback - the callback handling the response object
```

File: lib/IAMClient.js [hunk 5]
```diff
@@ -673,6 +673,38 @@ class VaultClient {
...
+     * and the error, if there is one
+     * @returns{undefined}
```
Comment (miniscruff) [inline lib/IAMClient.js]:
```suggestion
     * @returns {undefined}
```

File: tests/unit/getAccountIds.js (hunk 2 above)
Comment (jonathan-gramain) [reply tests/unit/getAccountIds.js:38]: OK, then what about passing nonsense in the query string to trigger an error that would be also triggered by the equivalent Vault request? We may also add a test case for a missing ID returning a valid response. That to keep the mock as close as possible to the Vault behavior.

Comment (tmacro) [reply tests/unit/getAccountIds.js:38]: I've change it to use the reqUid param. I don't think adding a test case for a missing accountId is needed as there is no branching in the code around this, and would be the same as a "valid" response just with different strings.

**Pull request merged**